### PR TITLE
Merge relayer config and database config

### PR DESCRIPTION
**Pull request by xitronix**

### CODE REVIEW

File: universal-login-relayer/lib/config/knexfile.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ function getTestKnexConfig(): KnexConfig {
     client: 'postgresql',
     connection: {
       database: 'universal_login_relayer_test',
-      user:     'postgres',
+      user: 'postgres',
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/config/knexfile.ts]: The purpose of this PR should be to get rid of `get*KnexConfig` at all. All data should be inlined.

File: universal-login-relayer/test/helpers/startRelayer.js [hunk 2]
```diff
@@ -35,7 +36,19 @@ async function startRelayer(wallet, relayerConstructor) {
...
+      country: 'any'
+    },
+    onRampProviders: {
+      safello: {
+        appId: '1234-5678',
+        baseAddress: 'https://app.s4f3.io/sdk/quickbuy.html',
+        addressHelper: true
+      }
+    },
+    database: getKnexConfig(),
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/startRelayer.js]: There is apparent duplication between `startRelayer` and `createPreconfiguredRelayer`. Can we get rid of it?

File: universal-login-relayer/lib/http/relayers/RelayerUnderTest.ts [hunk 3]
```diff
@@ -66,7 +67,18 @@ export class RelayerUnderTest extends Relayer {
           baseAddress: 'https://app.s4f3.io/sdk/quickbuy.html',
           addressHelper: true
         }
-      }
+      },
+      database: {
+        client: 'postgresql',
```
Comment (JustynaBroniszewska) [inline universal-login-relayer/lib/http/relayers/RelayerUnderTest.ts]: Why not use `getTestKnexConfig()` here?

File: universal-login-commons/lib/core/types/common.ts [hunk 4]
```diff
@@ -5,3 +5,5 @@ export type Predicate = (...args: any[]) => boolean;
 export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
 
 export type PartialRequired<T, TRequired extends keyof T> = Partial<T> & Pick<T, TRequired>;
+
+export type RecursivePartial<T> = {[P in keyof T]?: RecursivePartial<T[P]>};
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/types/common.ts]: `DeepPartial`

File: universal-login-commons/lib/core/utils/deepMerge.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+export function deepMerge <T, U>(destination: T, source: U): T & U {
+  const result = JSON.parse(JSON.stringify(destination));
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/deepMerge.ts]: extract `deepCopy`

File: universal-login-commons/lib/core/utils/deepMerge.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  for (const property in source) {
+     if (typeof source[property] === 'object' &&
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/deepMerge.ts]: extract `isObject` (and add tests)

File: universal-login-commons/test/core/utils/deepMerge.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    }
+  };
+
+  const obj2 = {
+    a1: 10,
+    b: {
+      c2: 20,
+    }
+  };
+
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/deepMerge.ts]: it('empty merged to non-empty')
it('non-empty merged to empty')
it('flat merged to flat')
it('3 levels')

File: universal-login-relayer/lib/core/utils/config.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+export function getConfig(environment: string = getNodeEnv()) {
+  switch (environment) {
+    case 'production':
+      return prodConfig;
+    case 'test':
+      return testConfig;
+    case 'development':
+    default:
+      return devConfig;
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/core/utils/config.ts]: throw exeception

File: universal-login-relayer/lib/core/utils/config.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import {getEnv} from '@universal-login/commons';
+import devConfig from '../../config/config.dev';
+import prodConfig from '../../config/config.prod';
+import testConfig from '../../config/config.test';
+
+export const getNodeEnv = () => getEnv('NODE_ENV', 'development');
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/core/utils/config.ts]: I would not include default

File: universal-login-relayer/lib/config/config.dev.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
+import path from 'path';
+import {utils} from 'ethers';
+import {getEnv, ETHER_NATIVE_TOKEN} from '@universal-login/commons';
+import {Config} from './relayer';
+
+
+const migrationDir = path.join(__dirname, '../integration/sql/migrations');
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/config/config.dev.ts]: inline

File: universal-login-relayer/lib/config/config.dev.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+export const config: Config =  Object.freeze({
+  jsonRpcUrl: getEnv('JSON_RPC_URL', ''),
```
Comment (marekkirejczyk) [inline universal-login-relayer/lib/config/config.dev.ts]: Use 'GENERATED' if will be overridden.

File: universal-login-relayer/test/helpers/knex.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import knex from 'knex';
+import {getConfig} from '../../lib';
+
+export const getTestKnexConfig = () => getConfig('test').database;
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/knex.ts]: Unnecessary

File: universal-login-relayer/test/helpers/knex.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+
+export const getKnex = () => knex(getConfig('test').database);
```
Comment (marekkirejczyk) [inline universal-login-relayer/test/helpers/knex.ts]: getTestKnex

File: universal-login-commons/lib/core/utils/deepMerge.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+      result[property] = result[property] || {};
+      result[property] = deepMerge(result[property], source[property]);
+    } else {
+      result[property] = source[property];
+    }
+  }
+  return result;
+}
+
+export function isNotNullObject<T>(object: T): boolean {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/deepMerge.ts]: `isNotNullObject` -> `isNullObject`

File: universal-login-commons/lib/core/utils/deepMerge.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  return typeof object === 'object' && object !== null;
+}
+
+export function deepCopy<T>(object: T) {
+  return JSON.parse(JSON.stringify(object));
+}
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/deepMerge.ts]: Already implemented, please deduplicate.

https://github.com/UniversalLogin/UniversalLoginSDK/blob/d73088b8ffb34141613de3b7c5fed992d20b23d2/universal-login-ops/src/utils/utils.js

**Pull request merged**